tracing/src/main/java/com/palantir/tracing/Tracer.java

```diff
@@ -239,14 +247,13 @@ public static void setSampler(TraceSampler sampler) {
 
     /** Returns the globally unique identifier for this thread's trace. */
     public static String getTraceId() {
-        return currentTrace.get().getTraceId();
+        return Preconditions.checkNotNull(currentTrace.get(), "There is no root span").getTraceId();
```

Comment:
I considered returning a newly generated ID when no trace is available, but that could potentially result in bad data if we initialize a new span assuming that traceId will be maintained.

Reply:
Possibly want to use the safelogging Preconditions here so these messages will definitely not get redacted?

https://github.com/palantir/safe-logging#preconditions

Reply:
I agree that would be ideal, but we don't currently depend on safe-logging. I'd prefer not to add deps in this change.

Reply:
should we expose a `public boolean hasTraceId()` to allow checking if tracing is currently enabled?

Once this PR goes through we can obviously clean up https://github.com/palantir/tritium/blob/develop/tritium-tracing/src/main/java/com/palantir/tritium/tracing/TracingInvocationEventHandler.java#L71 as `getAndClearTrace` will now do the right thing and clear state, but there may be cases where we want to know if there is a root trace ID already set.

Reply:
Added.

I searched for usages of getTraceId, and everything that I found called it after creating a new span, so it should be safe, though there may be cases that I'm not aware of. For what it's worth, anything that does throw here is also responsible for leaking traceIds.